server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -247,20 +246,29 @@ boolean dispatchRequest(final RestRequest request, final RestChannel channel, fi
         } else {
             // Get the map of matching handlers for a request, for the full set of HTTP methods.
             final Set<RestRequest.Method> validMethodSet = getValidHandlerMethodSet(request);
-            if (validMethodSet.size() > 0
-                && validMethodSet.contains(request.method()) == false
-                && request.method() != RestRequest.Method.OPTIONS) {
-                // If an alternative handler for an explicit path is registered to a
-                // different HTTP method than the one supplied - return a 405 Method
-                // Not Allowed error.
-                handleUnsupportedHttpMethod(request, channel, validMethodSet);
-                requestHandled = true;
-            } else if (validMethodSet.contains(request.method()) == false
-                && (request.method() == RestRequest.Method.OPTIONS)) {
-                handleOptionsRequest(request, channel, validMethodSet);
+
+            try {
+                // this will throws an exception if the method is invalid
+                final RestRequest.Method requestMethod = request.method();
+
+                if (validMethodSet.size() > 0
+                    && validMethodSet.contains(requestMethod) == false
+                    && requestMethod != RestRequest.Method.OPTIONS) {
+                    // If an alternative handler for an explicit path is registered to a
+                    // different HTTP method than the one supplied - return a 405 Method
+                    // Not Allowed error.
+                    handleUnsupportedHttpMethod(request, channel, validMethodSet, null);
+                    requestHandled = true;
+                } else if (validMethodSet.contains(requestMethod) == false
+                    && (requestMethod == RestRequest.Method.OPTIONS)) {
+                    handleOptionsRequest(request, channel, validMethodSet);
+                    requestHandled = true;
+                } else {
+                    requestHandled = false;
+                }
+            } catch (IllegalArgumentException e) {
+                handleUnsupportedHttpMethod(request, channel, validMethodSet, e);
```

Comment:
More of an optional question/suggestion: Can't we just do this a lot simpler by simply invoking this method upstream in `org.elasticsearch.rest.RestController#tryAllHandlers` even if it means that we have to duplicate the call to `getValidHandlerMethodSet`? (we'd save the whole change in this method and one level of indent)

Reply:
I pushed eac0baf906e04ec87f82d96224d1507a52b33c68, what do you think?

Reply:
Yea I like this a lot better :)

Reply:
Thanks!